library/pkcs11_client.c

```diff
@@ -138,6 +138,39 @@ static size_t pkcs11_pk_signature_size( const void *ctx_arg )
     }
 }
 
+static int pkcs11_sign_core( mbedtls_pk_pkcs11_context_t *ctx,
+                             CK_MECHANISM_TYPE mechanism_type,
+                             const unsigned char *payload, size_t payload_len,
+                             unsigned char *sig, size_t *sig_len,
+                             size_t sig_size )
+{
+    CK_ULONG ck_sig_len = sig_size;
+    CK_MECHANISM mechanism = {mechanism_type, NULL_PTR, 0};
+    CK_RV rv;
+    rv = C_SignInit( ctx->hSession, &mechanism, ctx->hPrivateKey );
+    if( rv != CKR_OK )
+        goto exit;
+    rv = C_Sign( ctx->hSession, (CK_BYTE_PTR) payload, payload_len,
+                 sig, &ck_sig_len );
+    if( rv != CKR_OK )
+        goto exit;
+    *sig_len = ck_sig_len;
+exit:
+    return( pkcs11_err_to_mbedtls_pk_err( rv ) );
```

Comment:
Unless there PKCS11 demands that function must wipe target buffers before returning a failure, it would be safer to clear `sig` in case of failure.

Reply:
There is no such requirement in PKCS#11. It will often turn out this way in practice, because copying the data out of some protected memory or communication buffer is the last step, but it's best for us not to assume this and to zero out `sig` on failure, or alternatively to zero it out at the very beginning of the function.